File: src/platform/posix/dbpath.c
/*
 * Copyright 2014-2016, Björn Ståhl
 * License: 3-Clause BSD, see COPYING file in arcan source repository.
 * Reference: http://arcan-fe.com
 */

#include <stdlib.h>
#include <stdint.h>
#include <stdbool.h>
#include <stdio.h>
#include <libgen.h>
#include <string.h>
#include <math.h>
#include <assert.h>
#include <ctype.h>

#include <sys/stat.h>
#include <sys/types.h>
#include <unistd.h>

#include <arcan_math.h>
#include <arcan_general.h>

char* platform_dbstore_path()
{
	char* out = NULL;
	char* homedir = getenv("HOME");
	size_t len;

	if (homedir && (len = strlen(homedir))){
		char dirbuf[len + sizeof("/.arcan")];
		snprintf(dirbuf, sizeof(dirbuf), "%s/.arcan", homedir);

/* ensure it exists, ignore EEXIST */
		mkdir(dirbuf, S_IRWXU);
		char fbuf[sizeof(dirbuf) + sizeof("/arcan.sqlite")];
		snprintf(fbuf, sizeof(fbuf), "%s/.arcan/arcan.sqlite", homedir);
		return strdup(fbuf);
	}

	return NULL;
}

File: src/platform/posix/dbpath.zig
// Zig port of posix/dbpath.c
// Returns the default path to arcan's sqlite database (~/.arcan/arcan.sqlite).

const std = @import("std");
const c = @import("posix");

export fn platform_dbstore_path() [*c]u8 {
    const homedir: [*c]const u8 = c.getenv("HOME") orelse return null;
    const len = c.strlen(homedir);
    if (len == 0) return null;
    const home = homedir[0..len];

    // Build ~/.arcan directory path and ensure it exists
    var dirbuf: [4096]u8 = undefined;
    const dir = std.fmt.bufPrintZ(&dirbuf, "{s}/.arcan", .{home}) catch return null;
    _ = c.mkdir(dir.ptr, 0o700);

    // Build full database path
    var fbuf: [4096]u8 = undefined;
    const path = std.fmt.bufPrintZ(&fbuf, "{s}/.arcan/arcan.sqlite", .{home}) catch return null;

    return c.strdup(path.ptr);
}
